Optimize the calculation speed of `InterRDF_s`

gitzhangch opened the pull request:
Refactored bin index calculation in the _single_frame function of InterRDF_s to optimize redundant np.histogram calls. 

<!-- 
Insert issue number that this PR fixes (if any) just after 'Fixes #'.
If this PR does not fix an existing issue, consider opening one or remove 'Fixes #' from the PR description.
-->
Fixes #5067

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 
- Refactored bin index calculation in `InterRDF_s._single_frame` to avoid multiple redundant `np.histogram` calls, improving performance.

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [x] Issue raised/referenced?
 - [x] Tests updated/added?
 - [x] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5073.org.readthedocs.build/en/5073/

<!-- readthedocs-preview mdanalysis end -->

gitzhangch committed:
Refactored bin index calculation in the _single_frame function of InterRDF_s to optimize redundant np.histogram calls. Additionally, divided by N in _conclude to make the computation consistent with InterRDF.

orbeckst reviewed (changes requested):
Thank you for the contribution! With any code rewrite for performance, the first thing is to make sure that the existing tests still pass. Have a look at the test output (you can run the tests locally, too) and modify your code so that it produces the same results as before.

If you have any reason to belief that the new code produces correct output while the old code is incorrect, please start a discussion.

gitzhangch committed:
Adjust the code to ensure the test passes.

gitzhangch committed:
Fewer revisions

gitzhangch committed:
formt revisions

gitzhangch commented:
@orbeckst 
During my work with `MDAnalysis.analysis.rdf.InterRDF_s`, I noticed that processing Gromacs trajectories was much slower compared to `InterRDF`. Upon investigation, I found that the `_single_frame` method was repeatedly calling `np.histogram` inside a loop, causing significant slowdowns. Additionally, after running `rdf_calc.run()`, plotting with `plt.plot(rdf.results.bins, rdf.results.rdf[0][0, 0])` showed that `rdf.results.rdf[0][0, 0]` was mostly zeros, indicating issues with the calculation logic. I also observed that `_conclude` did not normalize the arrays as in `InterRDF` (i.e., dividing by N = nA * nB), and that the `distances.capped_distance` call in `_single_frame` lacked the `min_cutoff` parameter, making it inconsistent with the `range` setting.

**Commits and testing:**
- **First commit:**  
  1. Removed repeated `np.histogram` calls to improve speed  
  2. Modified `self.results.count[i][0, 0]` to fix RDF accuracy  
  3. Updated `_conclude` to normalize by N  
  4. Set `min_cutoff` in `distances.capped_distance`  
  However, these changes caused many test failures.

- **Second commit:**  
  Removed changes 2 and 3. Most tests passed, but two still failed.

- **Third commit:**  
  Reverted change 4. Now, almost all tests pass except for a Codecov/coverage warning.

I don't use tests very often, so I've briefly described the test situation here. I hope to discuss with you how to make reasonable modifications further.

gitzhangch committed:
Merge branch 'MDAnalysis:develop' into develop

gitzhangch committed:
fix for test

gitzhangch committed:
fix for test

gitzhangch committed:
fix for test

gitzhangch committed:
fix for test

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,20 @@ def _single_frame(self):
                 backend=self.backend,
             )
 
+            # The following is an optimized version based on the old logic.
```
remove this comment because it's not useful for anyone reading the code (what is "old logic"?); if you want to communicate this information to reviewers, comment in GitHub on your code (basically, review yourself)

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,20 @@ def _single_frame(self):
+            bin_indices = (dist - minv) * bins / (maxv - minv)
+            bin_indices = bin_indices.astype(np.int64)
+
             for j, (idx1, idx2) in enumerate(pairs):
```
Perhaps there's even a way to replace the whole loop with a single numpy assignment, using [`np.isin()`](https://numpy.org/doc/stable/reference/generated/numpy.isin.html).

orbeckst replied:
(This is more of a comment than a request – if your code demonstrably speeds up the calculation while still producing correct results then I'd be happy to merge.)

orbeckst replied:
The more I am looking at this loop, the more I think that it should be completely eliminated. All it seems to be doing is to check if a distance associated with pair j is inside the bounds of the histogram.

Can you use [`np.digitize()`](https://numpy.org/doc/stable/reference/generated/numpy.digitize.html) to handle bin-indices together with all the range considerations? You may be able to replace your code above and the loop below.

gitzhangch replied:
The code has been updated, and the loop has been replaced.

orbeckst reviewed (changes requested):
Do you have some minimal performance data (before/after)?

Can you add an entry to CHANGELOG where you summarize your changes? It looks to me that the current version does not change normalization and simply addresses performance? (That's fine and makes it easier to merge the PR... if results change then that's more tricky.)

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,20 @@ def _single_frame(self):
-                count, _ = np.histogram(dist[j], **self.rdf_settings)
-                self.results.count[i][idx1, idx2, :] += count
+                self.results.count[i][idx1, idx2, bin_indices[j]] += int(
+                    bin_indices[j] in range(bins)
```
Definitely address this one. I ran a quick microbenchmark:
```python
In [1]: bins = 123

In [2]: x = 78.9

In [3]: y = 222.222

In [4]: range_object = range(bins)

In [5]: %timeit x in range_object
1.93 μs ± 48.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [6]: %timeit 0 <= x < bins
50.6 ns ± 0.248 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit y in range_object
1.88 μs ± 13.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [8]: %timeit 0 <= y < bins
51.5 ns ± 0.435 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

# Do not cache range object
In [9]: %timeit x in range(bins)
2.07 μs ± 84.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

The look-up from `range` is ~40 times SLOWER than the comparison operators. Definitely use the suggestion.
Even

orbeckst replied:
The problem is not so much the creation of the range object (see last line) but that `in range(...)` is slow.

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,20 @@ def _single_frame(self):
-                self.results.count[i][idx1, idx2, :] += count
+                self.results.count[i][idx1, idx2, bin_indices[j]] += int(
+                    bin_indices[j] in range(bins)
+                )
```
That's probably a minor thing but should be addressed for consistency.

orbeckst reviewed (changes requested):
See comments, in particular can you rewrite with np.digitize?

gitzhangch committed:
remove for loop

gitzhangch committed:
fix

gitzhangch commented:
> See comments, in particular can you rewrite with np.digitize?

@orbeckst 
np.digitize is more readable but slower; manual calculation is faster for equal-width binning.
```
In [1]: import numpy as np

In [2]: arr = np.random.uniform(0, 10, size=1_000_000)

In [3]: timeit np.digitize(arr, np.linspace(0, 10, 101), right=False) - 1
34.3 ms ± 945 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: timeit ((arr - 0) * 100 / (10 -0)).astype(np.int64)
3.37 ms ± 52.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

gitzhangch commented:
@orbeckst 
### performance 

| Version         | Time per 10 frame | Speedup |
|-----------------|---------------|---------|
| Original (for loop + np.histogram) | 27.1 s ± 55.7 ms         | 1x      |
| Optimized (NumPy vectorized)       | 158 ms ± 513 µs        | 170x  |

### test code
```py
>>> rdf1 = before(u, [[ag1, ag2]]) # alias InterRDF_s as before and after
>>> %timeit rdf1.run(start=0,stop=10)
27.1 s ± 55.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> rdf2 = after(u, [[ag1, ag2]])
>>> %timeit rdf2.run(start=0,stop=10)
158 ms ± 513 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

gitzhangch committed:
Update CHANGELOG

Added new contributors and optimized performance of RDF calculation.

orbeckst commented on package/CHANGELOG:
```diff
@@ -77,6 +78,9 @@ Enhancements
  * Enables parallelization for `analysis.polymer.PersistenceLength` (Issue #4671, PR #5074)
+ * Optimized performance of `analysis.rdf.InterRDF_s._single_frame` by 
+   replacing the inner for loop and repeated `np.histogram` calls with a 
+   vectorized NumPy calculation.(Issue #5067, PR #5073)
```
We don't need the deep technical details: the CHANGELOG is mainly for users to understand what changed for them so we can just write

```suggestion
 * Improved performance of `analysis.rdf.InterRDF_s` about 100-fold  
   by using vectorized NumPy calculations. (Issue #5067, PR #5073)
```

We also traditionally add the newest CHANGELOG entries _at the top_ of the section so just move the bullet point directly under *Enhancements*.

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,18 @@ def _single_frame(self):
+            bin_indices = (dist - minv) * bins / (maxv - minv)
+            bin_indices = bin_indices.astype(np.int64)
+            counts = np.isin(bin_indices, np.arange(bins)).astype(np.int64)
+            bin_indices = bin_indices * counts
```
This looks a bit like a hack. `isin()` returns a boolean array so wouldn't it be cleaner to pick out only the bin indices where we have a count
```python
bins_with_counts = np.isin(bin_indices, np.arange(bins))
selected_bin_indices = bin_indices[bins_with_counts]
...
self.results.count[i][idx1s, idx2s, bin_indices] += 1
```
Please check.

gitzhangch replied:
idx1s, idx2s, bin_indices should have the same shape. It's error to pick out only the bin indices where we have a count `isin()`. For example
```
In [7]: arr[a,b]+=1
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[7], line 1
----> 1 arr[a,b]+=1

IndexError: shape mismatch: indexing arrays could not be broadcast together with shapes (3,) (2,)
```
But you’ve given me inspiration. `bin_indices = bin_indices * counts` is unnecessary.

gitzhangch committed:
Improve performance of InterRDF_s with vectorized NumPy calculations and update CHANGELOG.

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,17 @@ def _single_frame(self):
-            for j, (idx1, idx2) in enumerate(pairs):
-                count, _ = np.histogram(dist[j], **self.rdf_settings)
-                self.results.count[i][idx1, idx2, :] += count
+            bins = self.rdf_settings["bins"]
```
Can we rename to `nbins` to make clear that these are simple equi-distant bins?
```suggestion
            nbins = self.rdf_settings["bins"]
```
and then change below.

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,17 @@ def _single_frame(self):
+            minv, maxv = (
+                self.rdf_settings["range"][0],
+                self.rdf_settings["range"][1],
+            )
```
We should be able to just use tuple assignment
```suggestion
           rmin, rmax = self.rdf_settings["range"]
```
I would also rename the variables. "v" does not relate to anything whereas "r" is really the distance in g(r).

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,17 @@ def _single_frame(self):
+                self.rdf_settings["range"][0],
+                self.rdf_settings["range"][1],
+            )
+            bin_indices = (dist - minv) * bins / (maxv - minv)
```
add a comment above explaining briefly that we are manually histogramming the distances for the equidistant bins and that this is faster than calling np.histogram() or np.digitize()

orbeckst commented on package/MDAnalysis/analysis/rdf.py:
```diff
@@ -645,9 +645,17 @@ def _single_frame(self):
+            )
+            bin_indices = (dist - minv) * bins / (maxv - minv)
+            bin_indices = bin_indices.astype(np.int64)
+            counts = np.isin(bin_indices, np.arange(bins)).astype(np.int64)
```
We're just checking for any counts inside the range of bins, right?

Would it be faster to just check that `0 <= bin_indices < nbins` (might have to write with `np.logical_and()`?

orbeckst replied:
Microbenchmark for 1000 indices over 75 bins:
```python
>>> import numpy as np
>>> nbins = 75
>>> bins = np.arange(nbins)
>>> rng = np.random.default_rng()
>>> ix = rng.integers(0, 120, 1000)
>>> %timeit (0 <= ix) & (ix < nbins)
1.63 μs ± 4.09 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

>>> %timeit np.isin(ix, bins)
16.2 μs ± 625 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# same result
>>> np.all(((0 <= ix) & (ix < nbins) ) == np.isin(ix, bins))
Out[16]: np.True_
```
It looks as if the simple comparison is ~10x faster than np.isin() and gives the same answer.

However, please let me know if I am overlooking anything. I am also aware that I originally proposed np.isin() — I haven't looked at the history but perhaps you had a comparison there already before.

orbeckst reviewed (changes requested):
* Can you please adjust variable names to make the code more readable?
* I think we can boost the speed further by replacing `np.isin()`.

See comments below. I think that's really it.

Can you do a quick benchmark of the final code version and post the result?

orbeckst commented:
We just merged https://github.com/MDAnalysis/mdanalysis/pull/4884 that enables parallelization. We do want to check that your optimized code also works in the parallel context.

orbeckst committed:
Merge branch 'develop' of https://github.com/gitzhangch/mdanalysis into gitzhangch-develop

orbeckst committed:
fixed change entries for PR #4884 on RDF parallelization

- corrected version changed dates for PR #4884 in analysis.rdf to 2.10.0
- moved entry in CHANGELOG to 2.10.0

orbeckst commented:
@gitzhangch when continuing on this PR, do a `git pull` first because I merged the latest develop and I also made some updates.

gitzhangch committed:
update code, comment and  variables

gitzhangch committed:
fix format

gitzhangch committed:
update format

orbeckst reviewed (approved):
Looks great. Thanks for addressing my last comments.

Have you done a quick before/after benchmark? I'd like to see the improvement before I'll merge.

gitzhangch commented:
### before/after benchmark
 @orbeckst 
 I performed a simple and quick parameter setup test. Since I don't use conda, the dev version of the package was installed in a separate test environment to benchmark the code separately before and after the update; for performance details, please refer to the following.  

Because the atomic pair size set in the benchmark is relatively small, the performance is not as striking as before. The larger the `natoms**2`, the greater the computational difference between the old and new code.  

Additionally, is it necessary to update the corresponding files in the benchmark folder?
#### code
```python
import MDAnalysis

try:
    from MDAnalysisTests.datafiles import TPR, XTC
except:
    pass

try:
    from MDAnalysis.analysis.rdf import InterRDF_s
except:
    pass


class SimpleRdfBench(object):
    """Benchmarks for MDAnalysis.analysis.rdf"""

    params = ([20, 75, 200], [[0, 5], [0, 15], [0, 20]], [10, 100], [1, 3, 9])

    param_names = ["nbins", "range_val", "natoms", "npairs"]

    def setup(self, nbins, range_val, natoms, npairs):

        self.sel_str = "name OW"

        self.u = MDAnalysis.Universe(TPR, XTC)

        try:
            self.sel = self.u.select_atoms(self.sel_str)[:natoms]
        except AttributeError:
            self.sel = self.u.selectAtoms(self.sel_str)[:natoms]

        ags = [[self.sel, self.sel]] * npairs
        self.rdfs = InterRDF_s(u=self.u, ags=ags, nbins=nbins, range=range_val)

    def time_interrdf_s(self, nbins, range_val, natoms, npairs):
        """Benchmark a full trajectory parse
        by MDAnalysis.analysis.rdf.InterRDF_s
        """
        self.rdfs.run(stop=5)
```

#### result
before
```
[100.00%] ··· ======= =========== ======== ============= ============= =============
              --                                             npairs
              ---------------------------- -----------------------------------------
               nbins   range_val   natoms        1             3             9
              ======= =========== ======== ============= ============= =============
                 20      [0, 5]      10     16.2±0.03ms    25.1±0.1ms    52.0±0.5ms
                 20      [0, 5]     100      54.4±0.3ms     141±1ms       399±5ms
                 20     [0, 15]      10     18.0±0.08ms   30.7±0.04ms    68.5±0.2ms
                 20     [0, 15]     100       218±1ms       632±2ms      1.86±0.01s
                 20     [0, 20]      10     20.3±0.04ms    37.7±0.1ms    90.2±0.7ms
                 20     [0, 20]     100       426±3ms       1.25±0s       3.74±0s
                 75      [0, 5]      10     16.2±0.01ms   25.3±0.07ms    52.8±0.4ms
                 75      [0, 5]     100       56.6±1ms      150±4ms       426±20ms
                 75     [0, 15]      10     18.1±0.03ms    30.9±0.1ms    69.1±0.2ms
                 75     [0, 15]     100      220±0.8ms      643±6ms      1.89±0.02s
                 75     [0, 20]      10      20.4±0.1ms    37.8±0.1ms    89.7±0.3ms
                 75     [0, 20]     100       433±3ms      1.27±0.01s    3.79±0.03s
                200      [0, 5]      10      16.4±0.2ms    25.8±0.2ms   53.5±0.06ms
                200      [0, 5]     100       60.1±1ms      161±7ms       461±10ms
                200     [0, 15]      10     18.2±0.07ms   31.4±0.08ms    70.5±0.3ms
                200     [0, 15]     100       232±2ms       677±7ms      1.99±0.03s
                200     [0, 20]      10      20.8±0.2ms    38.7±0.1ms    92.7±0.9ms
                200     [0, 20]     100       455±5ms      1.33±0.01s    3.99±0.06s
              ======= =========== ======== ============= ============= =============
```

after
```
[100.00%] ··· ======= =========== ======== ============= ============= =============
              --                                             npairs
              ---------------------------- -----------------------------------------
               nbins   range_val   natoms        1             3             9
              ======= =========== ======== ============= ============= =============
                 20      [0, 5]      10     12.3±0.09ms   14.0±0.08ms   19.0±0.07ms
                 20      [0, 5]     100      14.6±0.1ms    20.5±0.3ms    39.1±0.8ms
                 20     [0, 15]      10     12.5±0.08ms   14.2±0.07ms    19.5±0.1ms
                 20     [0, 15]     100      19.2±0.1ms    34.5±0.4ms     79.8±2ms
                 20     [0, 20]      10     12.3±0.04ms   13.7±0.05ms   18.0±0.06ms
                 20     [0, 20]     100      17.7±0.2ms    29.7±0.5ms     66.9±2ms
                 75      [0, 5]      10      12.5±0.2ms    14.2±0.2ms    19.3±0.1ms
                 75      [0, 5]     100      16.0±0.3ms    25.2±0.7ms     52.3±4ms
                 75     [0, 15]      10      12.5±0.2ms    14.4±0.2ms    19.6±0.2ms
                 75     [0, 15]     100      20.7±0.3ms     38.8±1ms      100±10ms
                 75     [0, 20]      10     12.3±0.07ms    13.8±0.1ms    18.2±0.2ms
                 75     [0, 20]     100      19.4±0.3ms    34.2±0.7ms    82.7±10ms
                200      [0, 5]      10      12.5±0.1ms    14.3±0.2ms    19.5±0.1ms
                200      [0, 5]     100     18.7±0.04ms    35.2±0.9ms     86.0±6ms
                200     [0, 15]      10      12.5±0.1ms    14.4±0.1ms    19.8±0.2ms
                200     [0, 15]     100      24.7±0.9ms     48.4±3ms      152±20ms
                200     [0, 20]      10      12.4±0.1ms   13.7±0.01ms   18.2±0.03ms
                200     [0, 20]     100      22.4±0.3ms    42.2±0.7ms     122±20ms
              ======= =========== ======== ============= ============= =============
```

gitzhangch committed:
solve CHANGELOG conflict

gitzhangch committed:
update CHANGELOG

gitzhangch committed:
Merge branch 'develop' into develop

orbeckst commented on package/CHANGELOG:
```diff
@@ -45,6 +45,9 @@ Fixes
    directly passing them. (Issue #3520, PR #5006)
 
 Enhancements
+ * Improved performance of `analysis.rdf.InterRDF_s` about 100-fold  
```
Given the benchmark data, let's be a bit more conservative:
```suggestion
 * Improved performance of `analysis.rdf.InterRDF_s` > 20-fold  
```

orbeckst commented on package/CHANGELOG:
```diff
@@ -16,8 +16,8 @@ The rules for this file:
-         hmacdope, tylerjereddy, cbouy, talagayev, DrDomenicoMarson, amruthesht,
-         p-j-smith
+         hmacdope, tylerjereddy, cbouy, talagayev, DrDomenicoMarson, amruthesht, 
+         p-j-smith, gitzhangch
```
Are you already in AUTHORS?

orbeckst reviewed (commented):
Could you please also add the timing benchmark to `benchmarks/analysis/rdf.py`? That's a great addition!

Please check that you're also in AUTHORS – I didn't look.

Thank you!

orbeckst committed:
more conservative speedup claim in CHANGELOG

orbeckst commented:
I toned down the claim of "100-fold" speedup to "> 20-fold", simply because that's what you demonstrate here. It may well be much faster for larger N and that's great but let's be conservative.

I really like that you created a new benchmark case for InterRDF_s and I _really_ like you to add it to the code base, please. It's really good practice, especially for anything performance-related.

I'll just wait with merging until you had a chance to look at the PR and I'll do it as soon as you're done here. Excellent work @gitzhangch !

orbeckst commented:
@gitzhangch in order to get these changes into the next release, would you be able to respond by tomorrow, Tuesday?

gitzhangch committed:
update AUTHORS and rdf benchmark

gitzhangch committed:
Merge branch 'develop' of github.com:gitzhangch/mdanalysis into develop

gitzhangch committed:
update benchmark for InterRDF_s

gitzhangch commented:
@orbeckst I update AUTHORS and rdf benchmark.

orbeckst commented:
Fantastic, will merge when CI is all green.

IAlibay committed:
Merge branch 'develop' into develop

orbeckst commented:
Thank you for your contribution @gitzhangch , it will be in release 2.10.0! 🎉
